packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
 			delete rest.spacing;
 			delete rest.updateStatesAndBounds;
 
+			const
+				{dataSize, direction, itemSize, overhang, rtl, spacing} = this.prevProps,
+				{props} = this,
+				hasMetricsChanged = (
+					direction !== props.direction ||
+					((itemSize instanceof Object) ? (itemSize.minWidth !== props.itemSize.minWidth || itemSize.minHeight !== props.itemSize.minHeight) : itemSize !== props.itemSize) ||
+					overhang !== props.overhang ||
+					spacing !== props.spacing
+				);
+
+			this.hasDataSizeChanged = (dataSize !== props.dataSize);
+
+			if (hasMetricsChanged || this.props.clientSize && this.primary === null && this.secondary === null) {
+				this.calculateMetrics(props);
+				this.updateStatesAndBounds(props, true);
```

Comment:
As the name of the function suggests, I wonder if we can translate this to `getDerivedStatesFromProps` so that we don't worry about directly mutating states.

This would require for `cbScrollTo` it to be called elsewhere (maybe in `componentDidUpdate`?) and I'm not sure if that would cause a too big of a change.

What do you think?

Reply:
>As the name of the function suggests, I wonder if we can translate this to `getDerivedStatesFromProps` so that we don't worry about directly mutating states.

As I mentioned in the JIRA, my first approach is the one that you mentioned. I moved all the code in `componentWillReceiveProps` to `getDrivedStateFromProps`.

As you can see the following code, I renamed `updateStateAndBounds` as `updateMetrics`, passed `prevProps`, `props`, and `prevState` to the function, then the function returned a new state. To do that, I needed to make all functions called from `updateStateAndBounds` returning a new state.
https://github.com/enactjs/enact/compare/feature/PLAT-72723-enhanced2#diff-decb3acaa813ab261efba3ee57c8b4f9R235

But I thought we needed to update lots of code to change member variables to state variables and member functions to class static functions or global functions because `getDrivedStateFromProps` was a static function and only props and states were passed.

If it is the right way, I'll keep going ahead instead of this PR.

Reply:
<On Feb 18th>
- The code above including `cbScrollTo` is moved to `componentDidUpdate`.